akka-projection-core/src/main/scala/akka/projection/Projection.scala

```diff
@@ -60,11 +67,31 @@ trait Projection[Envelope] {
 @InternalApi
 private[projection] object RunningProjection {
 
-  def withBackoff(source: () => Source[Done, _], settings: ProjectionSettings): Source[Done, _] =
+  def withBackoff(
+      source: () => Source[Done, _],
+      settings: ProjectionSettings,
+      projectionId: ProjectionId,
+      statusObserver: StatusObserver[_])(implicit ec: ExecutionContext): Source[Done, _] = {
+    val count = new AtomicInteger(0)
     RestartSource
       .onFailuresWithBackoff(settings.minBackoff, settings.maxBackoff, settings.randomFactor, settings.maxRestarts) {
-        () => source()
+        () =>
+          source()
+            .map { elem =>
+              // TODO can we do this in a more elegant/efficient way?
+              // reset on first element
+              count.set(0)
```

Comment:
I'm not sure why we want to reset the count here, but I played around with ways to get the first element using `wireTap` and a promise.

```scala
  val firstElement = Promise[Int]()
  val source: Source[Int, NotUsed] = Source(1 to 10)
  val result = source
    .wireTap(Flow[Int].take(1).to(Sink.foreach(firstElement.success)))
    .runWith(Sink.foreach(i => println(s"foreach: $i")))

  firstElement.future.foreach { i =>
    println(s"first element $i received!")
  }(system.dispatcher)
```

The `FlowMonitor` via `Flow.monitor` could be another option, but that would require a polling loop.

Reply:
I find that both approaches are spreading logic and will make it hard to reason about. 

This is all internal API, so I would change the API so something like: 

```scala
def withBackoff(
	source: StatusObserverWithState => Source[Done, _],
...
```

then further, we change the `mappedSource` signature to receive a `StatusObserverWithState`

The `StatusObserverWithState` would hold an instance of user define `StatusObserver` (as a delegate) and a counter. 

All that logic can move to inside the `mappedSource`. It will make it easier for use to follow the bits. 

It make it more complicated to implement alternative projections, but we can have some utility classes to help remove repetition.

Reply:
I'm leaning towards removing the count parameters and leave that up to the implementors of `StatusObserver`

Reply:
👍 
That would simplify things.

Reply:
counters removed